core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/KOptDescriptor.java

```diff
@@ -0,0 +1,271 @@
+     */
+    private final Integer[] p;
+
+    /**
+     * The order each node is visited when the tour is travelled in the successor direction. It the
+     * inverse of {@link KOptDescriptor#p} (i.e.
+     * {@link KOptDescriptor#p}[predecessorVisitingOrder[i]] == i
+     *
+     */
+    private final Integer[] q;
+
+    private final Integer[] incl;
+
+    private static Integer[] computeInEdges(Object[] t) {
+        Integer[] out = new Integer[t.length];
+        int k = (t.length - 1) >> 1;
+
+        out[1] = t.length - 1;
+        out[t.length - 1] = 1;
+        for (int i = 1; i < k; i++) {
+            out[2*i + 1] = 2 * i;
+            out[2*i] = 2*i + 1;
+        }
+
+        return out;
+    }
+
+    public KOptDescriptor(
+            Object[] t,
+            Function<Object, Object> SUC,
+            TriPredicate<Object, Object, Object> BETWEEN) {
+        this(t, computeInEdges(t), SUC, BETWEEN);
+    }
+
+    public KOptDescriptor(
+            Object[] t,
+            Integer[] incl,
+            Function<Object, Object> SUC,
+            TriPredicate<Object, Object, Object> BETWEEN) {
+        int i,j;
+        this.k = (t.length - 1) >> 1;
+        this.t = t;
+        this.p = new Integer[t.length];
+        this.q = new Integer[t.length];
+        this.incl = incl;
+
+        for (i = j = 1; j <= k; i += 2, j++) {
+            p[j] = (SUC.apply(t[i]) == t[i + 1]) ? i : i + 1;
+        }
+
+        Comparator<Integer> Compare = (pa, pb) -> pa.equals(pb)? 0 : (BETWEEN.test(t[p[1]], t[pa], t[pb])? -1 : 1);
+        Arrays.sort(p, 2, k + 1, Compare);
+
+        for (j = 2 * k; j >= 2; j -= 2) {
+            p[j - 1] = i = p[j / 2];
+            p[j] = ((i & 1) == 1)? i + 1 : i - 1;
+        }
+
+        for (i = 1; i <= 2 * k; i++) {
+            q[p[i]] = i;
+        }
+    }
+
+    public Integer[] getPermutation() {
+        return p;
+    }
+
+    /**
+     * Breaks the permutation p into its disjoint cycles
+     */
+    public KOptCycleInfo Cycles() {
+        int cycleCount = 0;
+        Integer[] indexToCycle = new Integer[p.length];
+        Set<Integer> remaining = IntStream.range(1, p.length).boxed().collect(Collectors.toSet());
+        while (!remaining.isEmpty()) {
+            Integer current = remaining.iterator().next();
+            remaining.remove(current);
+
+            while (true) {
+                indexToCycle[current] = cycleCount;
+                current = q[incl[p[current]]];
+                if (!remaining.contains(current)) {
+                    break;
+                }
+                remaining.remove(current);
+            }
+            cycleCount++;
+        }
+
+        return new KOptCycleInfo(cycleCount, indexToCycle);
+    }
+
+    public List<Pair<Object, Object>> getInEdges() {
+        List<Pair<Object, Object>> out = new ArrayList<>(2 * k);
+        for (int i = 1; i <= k; i++) {
+            out.add(Pair.of(t[2*i], t[(2*i + 1) % (2 * k)]));
+        }
+        return out;
+    }
```

Comment:
It appears that many computations in this class could be cached, as they depend on immutable data.